{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: reference_processing.html\n",
    "title: reference_processing\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp reference_processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "from collections import defaultdict\n",
    "from interlap import InterLap\n",
    "from joblib import Parallel, delayed\n",
    "from LAFITE.utils import Vividict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def gtf_line_split (entry):\n",
    "\t\"\"\"split gtf line\n",
    "\t\"\"\"\n",
    "\tlst = entry.rstrip().split('\\t')\n",
    "\tchrom, source, feature, start, end, score, strand, frame, attribute = lst\n",
    "\tstart = int(start)\n",
    "\tend = int(end)\n",
    "\n",
    "\treturn chrom, source, feature, start, end, score, strand, frame, attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def gtf2splicing(gtf, transcript_ref = 'transcript_id', gene_ref = 'gene_id', keepAttribute=False, mergeMode =False):\n",
    "\t\"\"\"preprocess the gtf file to gene and isoform level\n",
    "\t\"\"\"\n",
    "\tisoform_structure_dict = defaultdict(dict)\n",
    "\tgene_structure_dict = defaultdict(dict)\n",
    "\twith open (gtf) as f:\n",
    "\t\tfor line in f:\n",
    "\t\t\tif not line.startswith('#'):\n",
    "\t\t\t\tchrom, source, feature, start, end, score, strand, frame, attributes = gtf_line_split(line)\n",
    "\t\t\t\t\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tattributes = [a.strip() for a in attributes.strip(';').split('; ') if len(a)>0]\n",
    "\t\t\t\t\tif mergeMode:\n",
    "\t\t\t\t\t\tattributes = [a for a in attributes if any(i in a for i in [gene_ref, transcript_ref])]\n",
    "\t\t\t\t\tattributes = dict([a.replace('\"','').split(' ',1)[0:2] for a in attributes])\n",
    "\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\traise ValueError(\"Fatal: please check input GTF format!\\n\")\n",
    "\n",
    "\t\t\t\tif keepAttribute:\n",
    "\t\t\t\t\tif feature in ('transcript', 'mRNA'):\n",
    "\t\t\t\t\t\tgene_id = attributes[gene_ref]\n",
    "\t\t\t\t\t\ttranscript_id  = attributes[transcript_ref]\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand, transcript_id, gene_id)] = [attributes]\n",
    "\n",
    "\t\t\t\tif feature == \"exon\":\n",
    "\t\t\t\t\tgene_id = attributes[gene_ref]\n",
    "\t\t\t\t\ttranscript_id  = attributes[transcript_ref]\n",
    "\n",
    "\t\t\t\t\tif (chrom, strand, gene_id) in gene_structure_dict:\n",
    "\t\t\t\t\t\tgene_structure_dict[(chrom, strand, gene_id)].update([start,end])\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tgene_structure_dict[(chrom, strand, gene_id)] = set([start,end])\n",
    "\n",
    "\t\t\t\t\tif (chrom, strand, transcript_id, gene_id) in isoform_structure_dict:\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand, transcript_id, gene_id)].extend([start,end])\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand, transcript_id, gene_id)] = [start,end]\n",
    "\n",
    "\treturn gene_structure_dict, isoform_structure_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def gtf2splicing(gtf, transcript_ref = 'transcript_id', keepAttribute=False):\n",
    "\t\"\"\"preprocess the gtf file to gene and isoform level\n",
    "\t\"\"\"\n",
    "\tisoform_structure_dict = Vividict()\n",
    "\twith open (gtf) as f:\n",
    "\t\tfor line in f:\n",
    "\t\t\tif not line.startswith('#'):\n",
    "\t\t\t\tchrom, source, feature, start, end, score, strand, frame, attributes = gtf_line_split(line)\n",
    "\t\t\t\t\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tattributes = [a.strip().replace('\"','') for a in attributes.strip(';').split('\"; ') if len(a)>0]\n",
    "\t\t\t\t\tattributes = dict([a.split(' ',1)[0:2] for a in attributes])\n",
    "\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\traise ValueError(\"Fatal: please check input GTF format\\n\")\n",
    "\n",
    "\t\t\t\tif keepAttribute:\n",
    "\t\t\t\t\tif feature in ('transcript', 'mRNA'):\n",
    "\t\t\t\t\t\ttranscript_id  = attributes[transcript_ref]\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand)][transcript_id] = [attributes]\n",
    "\n",
    "\t\t\t\tif feature == \"exon\":\n",
    "\t\t\t\t\ttranscript_id  = attributes[transcript_ref]\n",
    "\t\t\t\t\tif transcript_id in isoform_structure_dict[(chrom, strand)]:\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand)][transcript_id].extend([start,end])\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tisoform_structure_dict[(chrom, strand)][transcript_id] = [start,end]\n",
    "\n",
    "\treturn isoform_structure_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# single thread\n",
    "\n",
    "class RefAnnotationExtraction:\n",
    "\t\"\"\"\n",
    "\textraction the splicing/exon information from the reference annotation\n",
    "\t\"\"\"\n",
    "\n",
    "\tdef __init__(self, ref_gtf):\n",
    "\t\tself.ref_gtf = ref_gtf\n",
    "\n",
    "\tdef annotation_extraction(self):\n",
    "\n",
    "\t\tref_exon = defaultdict(set) # reference exon start and end sites, the key will be chromosome, strand\n",
    "\t\tref_junction = defaultdict(set) # splicing junction from reference annotation\n",
    "\t\tref_mutple_exon_trans = Vividict() # splicing list for every chromosome, e.g. {'chr1,+':[[1,2,3,4,5],[111,223,44]], 'chr1,-':[[xxx,xxx,xxx],[xxx,xx,xxxx]]}\n",
    "\t\tref_single_exon_trans = defaultdict(set) # splicing list for every chromosome, but only for single exon transcript\n",
    "\t\tleft_sj_set = defaultdict(set)\n",
    "\t\tright_sj_set = defaultdict(set)\n",
    "\t\ttss_dict = defaultdict(set) #  transcript start sites set\n",
    "\n",
    "\t\tgene_structure_dict, ref_trans_structure_dict = gtf2splicing(self.ref_gtf)\n",
    "\n",
    "\t\tfor isoform, splicing_lst in ref_trans_structure_dict.items():\n",
    "\t\t\tchrom, strand, transcript_id, gene_id = isoform\n",
    "\t\t\tsplicing_lst.sort()\n",
    "\t\t\tread_splicing = splicing_lst[1:-1]\n",
    "\t\t\tstart, end = splicing_lst[0], splicing_lst[-1]\n",
    "\n",
    "\t\t\t# record the TSS\n",
    "\t\t\ttss_dict[(chrom, strand)].add(start) if strand == '+' else tss_dict[(chrom, strand)].add(end)\n",
    "\n",
    "\t\t\tif not read_splicing:\n",
    "\t\t\t\t# record single exon trans start and end site\n",
    "\t\t\t\tref_single_exon_trans[(chrom, strand)].add((start,end))\n",
    "\t\t\telse:\n",
    "\t\t\t\t# record the splicing site, splicing junction, exon for multi exon trans\n",
    "\t\t\t\tsplicing_lst = iter(splicing_lst)\n",
    "\t\t\t\tfor idx, exon in enumerate(zip(splicing_lst,splicing_lst)):\n",
    "\t\t\t\t\tref_exon[(chrom, strand)].add(exon)\n",
    "\t\t\t\t\tif idx > 0:\n",
    "\t\t\t\t\t\tright_sj = exon[0]\n",
    "\t\t\t\t\t\tleft_sj_set[(chrom, strand)].add(left_sj)\n",
    "\t\t\t\t\t\tright_sj_set[(chrom, strand)].add(right_sj)\n",
    "\t\t\t\t\t\tref_junction[(chrom, strand)].add((left_sj,right_sj))\n",
    "\t\t\t\t\tleft_sj = exon[1]\n",
    "\n",
    "\t\t\t\t# record the splicing structure and the tss and tes for multi exon trans\n",
    "\t\t\t\tif strand == \"-\":\n",
    "\t\t\t\t# \tread_splicing.reverse()\n",
    "\t\t\t\t\tse_site = [end, start]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tse_site = [start, end]\n",
    "\n",
    "\t\t\t\tref_mutple_exon_trans[(chrom, strand)][tuple(read_splicing)] = se_site\n",
    "\n",
    "\t\treturn ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict\n",
    "\n",
    "\tdef annotation_sorting(self):\n",
    "\n",
    "\t\tref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = self.annotation_extraction()\n",
    "\n",
    "\t\tfor i in tss_dict:\n",
    "\t\t\ttss_dict[i] = sorted(tss_dict[i])\n",
    "\n",
    "\t\tfor i in ref_mutple_exon_trans:\n",
    "\t\t\t# sort multi exon transcript by splicing junction number\n",
    "\t\t\ttmp_dict = {}\n",
    "\t\t\tfor k in sorted(ref_mutple_exon_trans[i], key=len, reverse=True):\n",
    "\t\t\t\ttmp_dict[k] = ref_mutple_exon_trans[i][k]\n",
    "\t\t\tref_mutple_exon_trans[i] = tmp_dict\n",
    "\n",
    "\t\tfor i in left_sj_set:\n",
    "\t\t\tleft_sj_set[i] = sorted(left_sj_set[i])\n",
    "\t\tfor i in right_sj_set:\n",
    "\t\t\tright_sj_set[i] = sorted(right_sj_set[i])\n",
    "\n",
    "\t\t# covert to interlap data format\n",
    "\t\tfor i in ref_single_exon_trans:\n",
    "\t\t\tt = list(ref_single_exon_trans[i])\n",
    "\t\t\tref_single_exon_trans[i] = InterLap()\n",
    "\t\t\tref_single_exon_trans[i].update(t)\n",
    "\n",
    "\t\tfor i in ref_exon:\n",
    "\t\t\tt = list(ref_exon[i])\n",
    "\t\t\tref_exon[i] = InterLap()\n",
    "\t\t\tref_exon[i].update(t)\n",
    "\n",
    "\t\treturn ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class RefAnnotationExtraction:\n",
    "\t\"\"\"\n",
    "\textraction the splicing/exon information from the reference annotation\n",
    "\t\"\"\"\n",
    "\n",
    "\tdef __init__(self, chrom, strand, chrand_ref_trans_structure_dict):\n",
    "\t\tself.chrom = chrom\n",
    "\t\tself.strand = strand\n",
    "\t\tself.chrand_ref_trans_structure_dict = chrand_ref_trans_structure_dict\n",
    "\n",
    "\tdef annotation_extraction(self):\n",
    "\n",
    "\t\tchrand_ref_exon = set() # reference exon start and end sites, the key will be chromosome, strand\n",
    "\t\tchrand_ref_junction = set() # splicing junction from reference annotation\n",
    "\t\tchrand_ref_mutple_exon_trans = defaultdict(dict) # splicing list for every chromosome, e.g. {'chr1,+':[[1,2,3,4,5],[111,223,44]], 'chr1,-':[[xxx,xxx,xxx],[xxx,xx,xxxx]]}\n",
    "\t\tchrand_ref_single_exon_trans = set() # splicing list for every chromosome, but only for single exon transcript\n",
    "\t\tchrand_left_sj_set = set()\n",
    "\t\tchrand_right_sj_set = set()\n",
    "\t\tchrand_tss_dict = set() #  transcript start sites set\n",
    "\n",
    "\t\tfor isoform, full_block in self.chrand_ref_trans_structure_dict.items():\n",
    "\t\t\tfull_block.sort()\n",
    "\t\t\tiso_splicing = full_block[1:-1]\n",
    "\t\t\tstart, end = full_block[0], full_block[-1]\n",
    "\n",
    "\t\t\t# record the TSS\n",
    "\t\t\tchrand_tss_dict.add(start) if self.strand == '+' else chrand_tss_dict.add(end)\n",
    "\n",
    "\t\t\tif not iso_splicing:\n",
    "\t\t\t\t# record single exon trans start and end site\n",
    "\t\t\t\tchrand_ref_single_exon_trans.add((start,end,isoform))\n",
    "\t\t\telse:\n",
    "\t\t\t\t# record the splicing site, splicing junction, exon for multi exon trans\n",
    "\t\t\t\tfull_block = iter(full_block)\n",
    "\t\t\t\tfor idx, exon in enumerate(zip(full_block,full_block)):\n",
    "\t\t\t\t\tchrand_ref_exon.add(exon)\n",
    "\t\t\t\t\tif idx > 0:\n",
    "\t\t\t\t\t\tright_sj = exon[0]\n",
    "\t\t\t\t\t\tchrand_left_sj_set.add(left_sj)\n",
    "\t\t\t\t\t\tchrand_right_sj_set.add(right_sj)\n",
    "\t\t\t\t\t\tchrand_ref_junction.add((left_sj,right_sj))\n",
    "\t\t\t\t\tleft_sj = exon[1]\n",
    "\n",
    "\t\t\t\t# record the splicing structure and the tss and tes for multi exon trans\n",
    "\t\t\t\tif self.strand == '-':\n",
    "\t\t\t\t\tse_site = [end, start,isoform]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tse_site = [start, end,isoform]\n",
    "\t\t\t\tchrand_ref_mutple_exon_trans[tuple(iso_splicing)] = se_site\n",
    "\n",
    "\t\treturn chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_tss_dict\n",
    "\n",
    "\tdef annotation_sorting(self):\n",
    "\n",
    "\t\tchrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_tss_dict = self.annotation_extraction()\n",
    "\n",
    "\t\tchrand_tss_dict = sorted(chrand_tss_dict)\n",
    "\t\t# sort multi exon transcript by splicing junction number\n",
    "\t\tchrand_ref_mutple_exon_trans = dict(sorted(chrand_ref_mutple_exon_trans.items(), key=lambda d: len(d[0]), reverse=True))\n",
    "\n",
    "\t\tchrand_left_sj_set = sorted(chrand_left_sj_set)\n",
    "\t\tchrand_right_sj_set = sorted(chrand_right_sj_set)\n",
    "\n",
    "\t\t# covert to interlap data format\n",
    "\t\tif chrand_ref_single_exon_trans:\n",
    "\t\t\tt = list(chrand_ref_single_exon_trans)\n",
    "\t\t\tchrand_ref_single_exon_trans = InterLap()\n",
    "\t\t\tchrand_ref_single_exon_trans.update(t)\n",
    "\n",
    "\t\tif chrand_ref_exon:\n",
    "\t\t\tt = list(chrand_ref_exon)\n",
    "\t\t\tchrand_ref_exon = InterLap()\n",
    "\t\t\tchrand_ref_exon.update(t)\n",
    "\n",
    "\t\treturn self.chrom, self.strand, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_tss_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class RefProcessWrapper:\n",
    "\tdef __init__(self, ref_gtf, thread):\n",
    "\t\tself.ref_gtf = ref_gtf\n",
    "\t\tself.thread = thread\n",
    "\t\n",
    "\tdef process(self):\n",
    "\t\tpreprocess_lst = []\n",
    "\t\tref_trans_structure_dict = gtf2splicing(self.ref_gtf)\n",
    "\t\tfor (chrom, strand), chrand_ref_trans_structure_dict in ref_trans_structure_dict.items():\n",
    "\t\t\tpreprocess_lst.append(RefAnnotationExtraction(chrom, strand, chrand_ref_trans_structure_dict))\n",
    "\t\twith Parallel(n_jobs = self.thread) as parallel:\n",
    "\t\t\tresults = parallel(delayed(lambda x:x.annotation_sorting())(job) for job in preprocess_lst)\n",
    "\n",
    "\t\treturn results\n",
    "\t\n",
    "\tdef result_collection(self):\n",
    "\t\tresults = self.process()\n",
    "\t\tref_exon = defaultdict(dict) # reference exon start and end sites, the key will be chromosome, strand\n",
    "\t\tref_junction = defaultdict(dict) # splicing junction from reference annotation\n",
    "\t\tref_mutple_exon_trans = Vividict() # splicing list for every chromosome, e.g. {'chr1,+':[[1,2,3,4,5],[111,223,44]], 'chr1,-':[[xxx,xxx,xxx],[xxx,xx,xxxx]]}\n",
    "\t\tref_single_exon_trans = defaultdict(dict) # splicing list for every chromosome, but only for single exon transcript\n",
    "\t\tleft_sj_set = defaultdict(dict)\n",
    "\t\tright_sj_set = defaultdict(dict)\n",
    "\t\ttss_dict = defaultdict(dict) #  transcript start sites set\n",
    "\n",
    "\t\tfor result in results:\n",
    "\t\t\tchrom, strand, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_tss_dict = result\n",
    "\t\t\t\n",
    "\t\t\ttss_dict[(chrom,strand)] = chrand_tss_dict\n",
    "\t\t\tif chrand_ref_exon:\n",
    "\t\t\t\tref_exon[(chrom,strand)] = chrand_ref_exon\n",
    "\t\t\tif chrand_ref_junction:\n",
    "\t\t\t\tref_junction[(chrom,strand)] = chrand_ref_junction\n",
    "\t\t\tif chrand_ref_mutple_exon_trans:\n",
    "\t\t\t\tref_mutple_exon_trans[(chrom,strand)] = chrand_ref_mutple_exon_trans\n",
    "\t\t\tif chrand_ref_single_exon_trans:\n",
    "\t\t\t\tref_single_exon_trans[(chrom,strand)] = chrand_ref_single_exon_trans\n",
    "\t\t\tif chrand_left_sj_set:\n",
    "\t\t\t\tleft_sj_set[(chrom,strand)] = chrand_left_sj_set\n",
    "\t\t\tif chrand_right_sj_set:\n",
    "\t\t\t\tright_sj_set[(chrom,strand)] = chrand_right_sj_set\n",
    "\n",
    "\t\treturn ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def short_reads_sj_import(sj_tab, left_sj_set, right_sj_set):\n",
    "\t\"\"\"import the splicing junctions detected from short reads data (STAR SJ_tab)\n",
    "\t\"\"\"\n",
    "\tfor file in sj_tab:\n",
    "\t\twith open (file) as f:\n",
    "\t\t\tfor line in f:\n",
    "\t\t\t\tline = line.strip('\\n').split(\"\\t\")\n",
    "\t\t\t\tchrom = line[0]\n",
    "\t\t\t\tif line[3] in ['+', '1']:\n",
    "\t\t\t\t\tstrand = '+'\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tstrand = '-'\n",
    "\t\t\t\tleft_sj = int(line[1])-1\n",
    "\t\t\t\tright_sj = int(line[2])+1\n",
    "\t\t\t\tleft_sj_set[(chrom, strand)].add(left_sj)\n",
    "\t\t\t\tright_sj_set[(chrom, strand)].add(right_sj)\n",
    "\n",
    "\tfor i in left_sj_set:\n",
    "\t\tleft_sj_set[i] = sorted(left_sj_set[i])\n",
    "\tfor i in right_sj_set:\n",
    "\t\tright_sj_set[i] = sorted(right_sj_set[i])\n",
    "\t\n",
    "\treturn left_sj_set, right_sj_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def cage_tss_import(cage_tss, tss_dict):\n",
    "\t\"\"\"import the splicing junctions detected from short reads data (STAR SJ_tab)\n",
    "\t\"\"\"\n",
    "\tfor file in cage_tss:\n",
    "\t\twith open (file) as f:\n",
    "\t\t\tfor line in f:\n",
    "\t\t\t\tline = line.strip('\\n').split(\"\\t\")\n",
    "\t\t\t\tchrom, strand = line[0], line[3]\n",
    "\t\t\t\ttss_center = round(abs(int(line[2])-int(line[1]))/2 + 1)\n",
    "\t\t\t\tif tss_center not in tss_dict[(chrom, strand)]:\n",
    "\t\t\t\t\ttss_dict[(chrom, strand)].add(tss_center)\n",
    "\n",
    "\tfor i in tss_dict:\n",
    "\t\ttss_dict[i] = sorted(tss_dict[i])\n",
    "\n",
    "\treturn tss_dict"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
