{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: collapsing.html\n",
    "title: read_collapsing\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp read_collapsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "from collections import defaultdict, Counter\n",
    "from dataclasses import dataclass, field\n",
    "from time import strftime\n",
    "from tqdm import tqdm\n",
    "from multiprocessing import Pool\n",
    "from LAFITE.utils import loc_distance, Vividict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "@ dataclass\n",
    "class RawAttributeCollection:\n",
    "\tname: str\n",
    "\tstart: int\n",
    "\tend: int\n",
    "\tfsm: bool = False\n",
    "\tmulti_exon: bool = True\n",
    "\tcorrect_site: list = field(default_factory=list)\n",
    "\tmerge_gap: list = field(default_factory=list)\n",
    "\tpolyaed: bool = False\n",
    "\tlowCredit_junction: dict = field(default_factory=dict)\n",
    "\tsplicing_tag: list = field(default_factory=list)\n",
    "\trss_dis: int = None\n",
    "\tres_dis: int = None\n",
    "\tcollapsed_ID: str = None\n",
    "\treference_id: str = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class ReadCorrectionColappse:\n",
    "\tdef __init__(self, chrom, strand, chrand_processed_read, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window, mis_intron_length, polya_dict, corExcept_dis=0):\n",
    "\t\tself.chrom = chrom\n",
    "\t\tself.strand = strand\n",
    "\t\tself.chrand_processed_read = chrand_processed_read\n",
    "\t\tself.chrand_ref_exon = chrand_ref_exon\n",
    "\t\tself.chrand_ref_junction = chrand_ref_junction\n",
    "\t\tself.chrand_ref_single_exon_trans = chrand_ref_single_exon_trans\n",
    "\t\tself.chrand_ref_mutple_exon_trans = chrand_ref_mutple_exon_trans\n",
    "\t\tself.chrand_left_sj_set = chrand_left_sj_set\n",
    "\t\tself.chrand_right_sj_set = chrand_right_sj_set\n",
    "\t\tself.chrand_junction_dict = chrand_junction_dict\n",
    "\t\tself.sj_correction_window = sj_correction_window\n",
    "\t\tself.mis_intron_length = mis_intron_length\n",
    "\t\tself.polya_dict = polya_dict\n",
    "\t\tself.corExcept_dis = corExcept_dis\n",
    "\n",
    "\tdef single_exon_read_collapse(self, read, single_exon_read):\n",
    "\t\t\"\"\"\n",
    "\t\tremove the single-exon reads overlaped with exon from reference multi-exon transcript and collapse\"\"\"\n",
    "\t\tstart, end = read\n",
    "\t\tif self.chrand_ref_exon:\n",
    "\t\t\toverlapped_ref_exon = tuple(self.chrand_ref_exon.find(read))\n",
    "\t\t\tif overlapped_ref_exon:\n",
    "\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\tif exon[0] <= start+self.sj_correction_window and exon[1] >= end-self.sj_correction_window:\n",
    "\t\t\t\t\t\tread = []\n",
    "\t\t\t\t\t\tbreak\n",
    "\t\tif read:\n",
    "\t\t\tcounter=Counter(range(start,end+1))\n",
    "\t\t\tfor i in counter:\n",
    "\t\t\t\tif i in single_exon_read:\n",
    "\t\t\t\t\tsingle_exon_read[i] += counter[i]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tsingle_exon_read[i] = counter[i]\n",
    "\n",
    "\t\treturn single_exon_read\n",
    "\n",
    "\tdef RTS_refrence_distance(self, start, end, read_splicing):\n",
    "\t\t\"\"\"calculate the distance between read start/end site to the reference transcript start/end site for FSM reads\n",
    "\n",
    "\t\tArgs:\n",
    "\t\t\tstrand (str): strand information\n",
    "\t\t\tstart (int): genomic start position regardless of strand\n",
    "\t\t\tend (int): genomic end position regardless of strand\n",
    "\t\t\tread_splicing (tuple): corrected read splicing\n",
    "\t\t\tchrand_ref_mutple_exon_trans (dict): reference multi-exon transcript\n",
    "\t\t\"\"\"\n",
    "\t\tif self.strand == '+':\n",
    "\t\t\trss_dis = abs(self.chrand_ref_mutple_exon_trans[read_splicing][0]-start)\n",
    "\t\t\tres_dis = abs(self.chrand_ref_mutple_exon_trans[read_splicing][1]-end)\n",
    "\t\telse:\n",
    "\t\t\trss_dis = abs(self.chrand_ref_mutple_exon_trans[read_splicing][0]-end)\n",
    "\t\t\tres_dis = abs(self.chrand_ref_mutple_exon_trans[read_splicing][1]-start)\n",
    "\n",
    "\t\treturn rss_dis, res_dis\n",
    "\n",
    "\tdef multi_exon_read_correction(self, read_id, full_block):\n",
    "\t\t\"\"\"\n",
    "\t\tsplicing junction correction and collaspsing for multi-exon read\"\"\"\n",
    "\n",
    "\t\traw_splicing= tuple(full_block[1:-1])\n",
    "\t\traw_read_attribute = RawAttributeCollection(read_id, full_block[0], full_block[-1])\n",
    "\t\tcorrected_read_splicing = []\n",
    "\n",
    "\t\t# polya event checking\n",
    "\t\ttry:\n",
    "\t\t\tif self.polya_dict[read_id]: raw_read_attribute.polyaed = True\n",
    "\t\texcept:\n",
    "\t\t\tpass\n",
    "\t\t\n",
    "\t\tif raw_splicing in self.chrand_ref_mutple_exon_trans: # raw read_splicing matching with the reference\n",
    "\t\t\traw_read_attribute.fsm = True\n",
    "\t\t\traw_read_attribute.splicing_tag = 'FSM'\n",
    "\t\t\tcorrected_read_splicing = raw_splicing\n",
    "\t\t\traw_read_attribute.rss_dis, raw_read_attribute.res_dis = self.RTS_refrence_distance(raw_read_attribute.start, raw_read_attribute.end, corrected_read_splicing)\n",
    "\n",
    "\t\telse: # splicing site correction\n",
    "\t\t\titered_raw_splicing = iter(raw_splicing)\n",
    "\t\t\tfor idx, (left_sj, right_sj) in enumerate(zip(itered_raw_splicing, itered_raw_splicing)):\n",
    "\n",
    "\t\t\t\t# check splicing coverage and motif in raw data\n",
    "\t\t\t\tsj_pos = full_block.index(left_sj)\n",
    "\t\t\t\ttmp_sj = (self.chrom, self.strand, left_sj, right_sj)\n",
    "\t\t\t\tjunction_coverage = self.chrand_junction_dict[tmp_sj][0]\n",
    "\t\t\t\tjunction_motif = self.chrand_junction_dict[tmp_sj][2]\n",
    "\n",
    "\t\t\t\tif left_sj in self.chrand_left_sj_set and right_sj in self.chrand_right_sj_set:\n",
    "\t\t\t\t\tif (left_sj, right_sj) in self.chrand_ref_junction:\n",
    "\t\t\t\t\t\traw_read_attribute.splicing_tag.append('M')\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\traw_read_attribute.splicing_tag.append('KC')\n",
    "\n",
    "\t\t\t\telif self.chrand_left_sj_set:\n",
    "\t\t\t\t\tleft_dis, left_ref_sj = loc_distance(self.chrand_left_sj_set, left_sj)\n",
    "\t\t\t\t\tright_dis, right_ref_sj = loc_distance(self.chrand_right_sj_set, right_sj)\n",
    "\n",
    "\t\t\t\t\t# do not correct the splicing site once the edit distance > sj_correction_window\n",
    "\t\t\t\t\tif left_dis > self.sj_correction_window: left_ref_sj = left_sj\n",
    "\t\t\t\t\tif right_dis > self.sj_correction_window: right_ref_sj = right_sj\n",
    "\n",
    "\t\t\t\t\t# correction exception, splicing junction with edit distance less than the given value for both sides\n",
    "\t\t\t\t\tif self.corExcept_dis and left_dis <= self.corExcept_dis and right_dis <= self.corExcept_dis and junction_coverage > 1 and junction_motif == 'canonical':\n",
    "\t\t\t\t\t\traw_read_attribute.splicing_tag.append('EXC')\n",
    "\n",
    "\t\t\t\t\t# splicing site correction\n",
    "\t\t\t\t\telif left_dis <= self.sj_correction_window or right_dis <= self.sj_correction_window:\n",
    "\t\t\t\t\t\tif [left_sj, right_sj] == [left_ref_sj, right_ref_sj]:\n",
    "\t\t\t\t\t\t\tpass\n",
    "\t\t\t\t\t\telif full_block[sj_pos-1] < left_ref_sj < right_ref_sj < full_block[sj_pos+2]:\n",
    "\t\t\t\t\t\t\traw_read_attribute.correct_site.append([left_sj, right_sj])\n",
    "\t\t\t\t\t\t\tleft_sj, right_sj = left_ref_sj, right_ref_sj\n",
    "\n",
    "\t\t\t\t\t\tif left_sj in self.chrand_left_sj_set and right_sj in self.chrand_right_sj_set:\n",
    "\t\t\t\t\t\t\tif (left_sj, right_sj) in self.chrand_ref_junction:\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.splicing_tag.append('CM')\n",
    "\t\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.splicing_tag.append('CKC')\n",
    "\n",
    "\t\t\t\t#checking unintended small intron overlap with reference exons\n",
    "\t\t\t\tif len(raw_read_attribute.splicing_tag) == idx and right_sj - left_sj <= self.mis_intron_length:\n",
    "\t\t\t\t\tif self.chrand_ref_exon:\n",
    "\t\t\t\t\t\toverlapped_ref_exon = tuple(self.chrand_ref_exon.find((left_sj, right_sj)))\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\toverlapped_ref_exon = ()\n",
    "\t\t\t\t\t# compare the unintended intron with the ref_exon from multi-exon transcripts\n",
    "\t\t\t\t\tif idx == 0 and overlapped_ref_exon:\n",
    "\t\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\tif (len(full_block) == 4 and exon[1] >= right_sj and exon[0] <= left_sj):\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.merge_gap.append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.multi_exon = False\n",
    "\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\telif (exon[1] == full_block[sj_pos+2] and exon[0] <= left_sj):\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.merge_gap.append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\tbreak\n",
    "\n",
    "\t\t\t\t\telif idx == int(len(raw_splicing)/2 - 1) and overlapped_ref_exon:\n",
    "\t\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\tif exon[1] >= right_sj and exon[0] == full_block[sj_pos-1]:\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.merge_gap.append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\n",
    "\t\t\t\t\telif 0 < idx < int(len(raw_splicing)/2 -1) and overlapped_ref_exon:\n",
    "\t\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\tif exon[1] == full_block[sj_pos+2] and exon[0] == full_block[sj_pos-1]:\n",
    "\t\t\t\t\t\t\t\traw_read_attribute.merge_gap.append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\tbreak\n",
    "\n",
    "\t\t\t\t\t# compare the unintended intron with the exon from single-exon transcripts\n",
    "\t\t\t\t\telif not overlapped_ref_exon and len(full_block) == 4:\n",
    "\t\t\t\t\t\tif self.chrand_ref_single_exon_trans:\n",
    "\t\t\t\t\t\t\toverlapped_ref_exon = tuple(self.chrand_ref_single_exon_trans.find((left_sj, right_sj)))\n",
    "\n",
    "\t\t\t\t\t\tif overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\t\tif exon[1] >= right_sj and exon[0] <= left_sj:\n",
    "\t\t\t\t\t\t\t\t\traw_read_attribute.merge_gap.append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\t\traw_read_attribute.multi_exon = False\n",
    "\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\n",
    "\t\t\t\tcorrected_read_splicing.extend([left_sj, right_sj])\n",
    "\n",
    "\t\t\t\tif [full_block[sj_pos], full_block[sj_pos+1]] in raw_read_attribute.merge_gap:\n",
    "\t\t\t\t\traw_read_attribute.splicing_tag.append('UI')\n",
    "\t\t\t\t\tdel corrected_read_splicing[-2:]\n",
    "\n",
    "\t\t\t\telif len(raw_read_attribute.splicing_tag) == idx:\n",
    "\t\t\t\t\traw_read_attribute.splicing_tag.append('NC')\n",
    "\t\t\t\t\tif junction_coverage == 1:\n",
    "\t\t\t\t\t\traw_read_attribute.lowCredit_junction[idx+1] = junction_motif\n",
    "\n",
    "\t\t\tcorrected_read_splicing = tuple(corrected_read_splicing)\n",
    "\t\t\tif corrected_read_splicing in self.chrand_ref_mutple_exon_trans:\n",
    "\t\t\t\traw_read_attribute.fsm = True\n",
    "\t\t\t\traw_read_attribute.rss_dis, raw_read_attribute.res_dis = self.RTS_refrence_distance(raw_read_attribute.start, raw_read_attribute.end, corrected_read_splicing)\n",
    "\t\t\t\traw_read_attribute.reference_id = self.chrand_ref_mutple_exon_trans[corrected_read_splicing][2]\n",
    "\n",
    "\t\treturn corrected_read_splicing, raw_read_attribute\n",
    "\n",
    "\tdef multi_exon_read_collapse(self, corrected_read_splicing, raw_read_attribute, rss_dis_lst, res_dis_lst, multi_exon_read, collapsed_idx):\n",
    "\t\t\"\"\"collapsing multi-exon read\n",
    "\t\t\"\"\"\n",
    "\t\tprefix = 'POS' if self.strand == '+' else 'NEG'\n",
    "\t\tif raw_read_attribute.lowCredit_junction:\n",
    "\t\t\tpass\n",
    "\t\telif corrected_read_splicing:\n",
    "\t\t\tif corrected_read_splicing not in multi_exon_read:\n",
    "\t\t\t\tcollapsed_idx += 1\n",
    "\t\t\t\traw_read_attribute.collapsed_ID = f'{self.chrom}_{prefix}.{collapsed_idx}'\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing] = [[raw_read_attribute.start], [raw_read_attribute.end], [raw_read_attribute.polyaed], 1, raw_read_attribute.fsm, raw_read_attribute.collapsed_ID, [raw_read_attribute.reference_id]]\n",
    "\t\t\telse:\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing][0].insert(0, raw_read_attribute.start)\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing][1].insert(0, raw_read_attribute.end)\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing][2].insert(0,raw_read_attribute.polyaed)\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing][3] += 1\n",
    "    \t\t\t\n",
    "\t\t\t\traw_read_attribute.collapsed_ID = multi_exon_read[corrected_read_splicing][5]\n",
    "\t\t\t\tmulti_exon_read[corrected_read_splicing][6].insert(0,raw_read_attribute.reference_id)\n",
    "\n",
    "\t\tif raw_read_attribute.rss_dis:\n",
    "\t\t\trss_dis_lst.append(raw_read_attribute.rss_dis)\n",
    "\t\t\tres_dis_lst.append(raw_read_attribute.res_dis)\n",
    "\t\t# print(len(multi_exon_read[corrected_read_splicing]))\n",
    "\t\treturn multi_exon_read, rss_dis_lst, res_dis_lst, collapsed_idx, raw_read_attribute\n",
    "\n",
    "\tdef coco_operation (self, collapsed_idx = 0):\n",
    "\t\t\"\"\"main function for correcting splicing junction and collpasing reads\n",
    "\t\t\"\"\"\n",
    "\t\tcorrected_read = defaultdict(dict)\n",
    "\t\tcorrection_log = defaultdict(dict)\n",
    "\t\tsingle_exon_read = defaultdict(dict)\n",
    "\t\tmulti_exon_read = defaultdict(dict)\n",
    "\t\trss_dis_lst = []\n",
    "\t\tres_dis_lst = []\n",
    "\n",
    "\t\tfor read_id, full_block in tqdm(self.chrand_processed_read.items(), desc = f'{strftime(\"%Y-%m-%d %H:%M:%S\")}: Collapsing raw reads from {self.chrom} {self.strand}'):\n",
    "\n",
    "\t\t\t# single-exon read collapsing\n",
    "\t\t\tif len(full_block) == 2: \n",
    "\t\t\t\tsingle_exon_read = self.single_exon_read_collapse(full_block, single_exon_read)\n",
    "\t\t\t\tcorrected_read[read_id] = full_block\n",
    "\t\t\t# multi-exon read correction and collapsing\n",
    "\t\t\telse:\n",
    "\t\t\t\tcorrected_read_splicing, raw_read_attribute = self.multi_exon_read_correction(read_id, full_block)\n",
    "\t\t\t\tmulti_exon_read, rss_dis_lst, res_dis_lst, collapsed_idx, raw_read_attribute = self.multi_exon_read_collapse(corrected_read_splicing, raw_read_attribute, rss_dis_lst, res_dis_lst, multi_exon_read, collapsed_idx)\n",
    "\n",
    "\t\t\t\tif corrected_read_splicing:\n",
    "\t\t\t\t\tcorrected_read[read_id] = [raw_read_attribute.start, *list(corrected_read_splicing), raw_read_attribute.end]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tcorrected_read[read_id] = [raw_read_attribute.start, raw_read_attribute.end]\n",
    "\t\t\t\tcorrection_log[read_id] = raw_read_attribute\n",
    "\n",
    "\t\treturn self.chrom, self.strand, single_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = '/expt/zjzace/Nanopore_subcellular/Reference/gencode.v38.primary_assembly.annotation.sorted.gtf'\n",
    "ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefProcessWrapper(gtf, 16).result_collection()\n",
    "bed = '/expt/zjzace/Nanopore_subcellular/Analysis/Assembly/LAFITE/GLRA_tmp/A549_Cyto_LAFITE_tmp/bam.bed'\n",
    "fa = '/NFS/mnemosyne2/expt/zjzace/GenomeRef/GRCh38.primary_assembly.genome.fa'\n",
    "junction_dict, processed_read = read_grouping(bed, fa)\n",
    "polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/Analysis/Nanopolish/A549_Cyto_PolyA.res')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-04-13 12:04:51: Collapsing raw reads from chr19 -: 100%|██████████| 48521/48521 [00:05<00:00, 9070.96it/s] \n"
     ]
    }
   ],
   "source": [
    "\n",
    "chrom, strand = 'chr19', '-'\n",
    "chrand_processed_read = processed_read[(chrom, strand)]\n",
    "chrand_ref_exon = ref_exon[(chrom, strand)]\n",
    "chrand_ref_junction = ref_junction[(chrom, strand)]\n",
    "chrand_ref_single_exon_trans = ref_single_exon_trans[(chrom, strand)]\n",
    "chrand_ref_mutple_exon_trans = ref_mutple_exon_trans[(chrom,strand)]\n",
    "chrand_left_sj_set = left_sj_set[(chrom,strand)]\n",
    "chrand_right_sj_set = right_sj_set[(chrom,strand)]\n",
    "chrand_junction_dict = junction_dict[(chrom,strand)]\n",
    "sj_correction_window = 40\n",
    "mis_intron_length = 150\n",
    "corExcept_dis=4\n",
    "polya_dict=polya_dict\n",
    "chrom, strand, single_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst = ReadCorrectionColappse(chrom, strand, chrand_processed_read, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window, mis_intron_length, polya_dict, corExcept_dis).coco_operation()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class CoCoWrapper:\n",
    "\tdef __init__(self, thread, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, sj_correction_window, polya_dict, mis_intron_length, tmp_dir,corExcept_dis=0):\n",
    "\t\tself.thread = thread\n",
    "\t\tself.processed_read = processed_read\n",
    "\t\tself.ref_exon = ref_exon\n",
    "\t\tself.ref_junction = ref_junction\n",
    "\t\tself.ref_single_exon_trans = ref_single_exon_trans\n",
    "\t\tself.ref_mutple_exon_trans = ref_mutple_exon_trans\n",
    "\t\tself.left_sj_set = left_sj_set\n",
    "\t\tself.right_sj_set = right_sj_set\n",
    "\t\tself.junction_dict = junction_dict\n",
    "\t\tself.sj_correction_window = sj_correction_window\n",
    "\t\tself.polya_dict = polya_dict\n",
    "\t\tself.mis_intron_length = mis_intron_length\n",
    "\t\tself.tmp_dir = tmp_dir\n",
    "\t\tself.corExcept_dis = corExcept_dis\n",
    "\t\t\n",
    "\n",
    "\tdef job_compute(self):\n",
    "\n",
    "\t\tjob = []\n",
    "\t\tfor branch in self.processed_read:\n",
    "\t\t\tchrom, strand = branch\n",
    "\t\t\tchrand_processed_read = self.processed_read[branch]\n",
    "\t\t\tchrand_ref_exon = self.ref_exon[branch]\n",
    "\t\t\tchrand_ref_junction = self.ref_junction[branch]\n",
    "\t\t\tchrand_ref_single_exon_trans = self.ref_single_exon_trans[branch]\n",
    "\t\t\tchrand_ref_mutple_exon_trans = self.ref_mutple_exon_trans[branch]\n",
    "\t\t\tchrand_left_sj_set = self.left_sj_set[branch]\n",
    "\t\t\tchrand_right_sj_set = self.right_sj_set[branch]\n",
    "\t\t\tchrand_junction_dict = self.junction_dict[branch]\n",
    "\t\t\tjob.append(ReadCorrectionColappse(chrom, strand, chrand_processed_read, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans,\n",
    "                                     chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, self.sj_correction_window, self.mis_intron_length, self.polya_dict, self.corExcept_dis))\n",
    "\t\tp = Pool(processes = self.thread)\n",
    "\t\tresult = [p.apply_async(i.coco_operation, args=()) for i in job]\n",
    "\t\tp.close()\n",
    "\t\tp.join()\n",
    "\n",
    "\t\treturn result\n",
    "\n",
    "\tdef result_collection(self):\n",
    "\t\tcollected_single_exon_read = Vividict()\n",
    "\t\tcollected_multi_exon_read = Vividict()\n",
    "\t\tcollected_rss = []\n",
    "\t\tcollected_res = []\n",
    "\t\tpath_to_log = f'{self.tmp_dir}/read_correction.log'\n",
    "\t\tpath_to_corrected_bed = f'{self.tmp_dir}/Corrected_reads.bed'\n",
    "\n",
    "\t\tresult = self.job_compute()\n",
    "\t\twith open(path_to_log, 'w') as flog, open(path_to_corrected_bed, 'w') as fbed:\n",
    "\t\t\tfor res in result:\n",
    "\t\t\t\tchrom, strand, single_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst = res.get()\n",
    "\t\t\t\t\n",
    "\t\t\t\tcollected_single_exon_read[(chrom, strand)] = single_exon_read\n",
    "\t\t\t\tcollected_multi_exon_read[(chrom, strand)] = multi_exon_read\n",
    "\t\t\t\tcollected_rss.extend(rss_dis_lst)\n",
    "\t\t\t\tcollected_res.extend(res_dis_lst)\n",
    "\t\t\t\tfor read_id, raw_read_attribute in correction_log.items():\n",
    "\t\t\t\t\traw_read_attribute.name = raw_read_attribute.name.split('_', 1)[1]\n",
    "\t\t\t\t\tattributes = '\\t'.join('{}: {}'.format(key, value) for key, value in raw_read_attribute.__dict__.items())\n",
    "\t\t\t\t\tflog.write(f'{attributes}\\n')\n",
    "\t\t\t\t\n",
    "\t\t\t\tfor read_id, full_block in corrected_read.items():\n",
    "\t\t\t\t\tread_name = read_id.split('_', 1)[1]\n",
    "\t\t\t\t\tbed_block = splicing_to_bed_block(chrom, strand, read_name, full_block)\n",
    "\t\t\t\t\tfbed.write(f'{bed_block}\\n')\n",
    "\t\treturn collected_single_exon_read, collected_multi_exon_read, collected_rss, collected_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV4 +: 100%|██████████| 2661/2661 [00:00<00:00, 91686.87it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV5 -: 100%|██████████| 590/590 [00:00<00:00, 12759.49it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV6 -: 100%|██████████| 1364/1364 [00:00<00:00, 10556.48it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV2 +: 100%|██████████| 4088/4088 [00:00<00:00, 6357.57it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV1 +: 100%|██████████| 3615/3615 [00:01<00:00, 2790.63it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV2 -: 100%|██████████| 6835/6835 [00:01<00:00, 5238.81it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV4 -: 100%|██████████| 7002/7002 [00:01<00:00, 5559.16it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV1 -: 100%|██████████| 7764/7764 [00:01<00:00, 5626.42it/s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV7 -: 100%|██████████| 7413/7413 [00:01<00:00, 6001.32it/s]s]\n",
      "2022-04-12 19:18:26: Collapsing raw reads from SIRV3 -: 100%|██████████| 2899/2899 [00:01<00:00, 1956.22it/s]s]\n",
      "2022-04-12 19:18:28: Collapsing raw reads from SIRV3 +: 100%|██████████| 10834/10834 [00:00<00:00, 79545.93it/s]\n",
      "2022-04-12 19:18:27: Collapsing raw reads from SIRV5 +: 100%|██████████| 20402/20402 [00:00<00:00, 29591.10it/s]\n",
      "2022-04-12 19:18:28: Collapsing raw reads from SIRV6 +: 100%|██████████| 32687/32687 [00:01<00:00, 30373.87it/s]\n"
     ]
    }
   ],
   "source": [
    "from LAFITE.reference_processing import RefProcessWrapper, short_reads_sj_import\n",
    "from LAFITE.preprocessing import read_grouping, polya_signal_import, PolyAFinder\n",
    "from LAFITE.utils import temp_dir_creation, bam2bed, keep_tmp_file\n",
    "gtf = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'\n",
    "bed = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.sorted.bed'\n",
    "fa = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta'\n",
    "junction_dict, processed_read = read_grouping(bed, fa)\n",
    "ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefProcessWrapper(gtf, 16).result_collection()\n",
    "\n",
    "# polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.polya.res')\n",
    "polya_dict = PolyAFinder(processed_read, fa, '/home/zjzace/software/SQANTI3-4.1/data/polyA_motifs/mouse_and_human.polyA_motif.txt').polya_estimation()\n",
    "collected_single_exon_read, collected_multi_exon_read, tss, tes = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, sj_correction_window=40, polya_dict=polya_dict, mis_intron_length = 150, tmp_dir='.',corExcept_dis=4).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def single_exon_read_collapse(read, chrand_ref_exon, overhang, single_exon_read):\n",
    "\t\"\"\"remove the single-exon reads overlaped with exon from reference multi-exon transcript and collapse\n",
    "\n",
    "\tArgs:\n",
    "\t\tread (list): start and end position of the single-exon read (1 base, [1,100])\n",
    "\t\tchrand_ref_exon (interlap data): exon from reference multi-exon transcript\n",
    "\t\toverhang (int): tolerance distance\n",
    "\t\tsingle_exon_read (dict): returned collapsed single-exon reads\n",
    "\t\"\"\"\n",
    "\t\n",
    "\tstart, end = read\n",
    "\tif chrand_ref_exon:\n",
    "\t\toverlapped_ref_exon = tuple(chrand_ref_exon.find(read))\n",
    "\t\tif overlapped_ref_exon:\n",
    "\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\tif exon[0] <= start+overhang and exon[1] >= end-overhang:\n",
    "\t\t\t\t\tread = []\n",
    "\t\t\t\t\tbreak\n",
    "\tif read:\n",
    "\t\tcounter=Counter(range(start,end+1))\n",
    "\t\tfor i in counter:\n",
    "\t\t\tif i in single_exon_read:\n",
    "\t\t\t\tsingle_exon_read[i] += counter[i]\n",
    "\t\t\telse:\n",
    "\t\t\t\tsingle_exon_read[i] = counter[i]\n",
    "\t\n",
    "\treturn single_exon_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def RTS_refrence_distance(strand, start, end, read_splicing, chrand_ref_mutple_exon_trans):\n",
    "\t\"\"\"calculate the distance between read start/end site to the reference transcript start/end site for FSM reads\n",
    "\n",
    "\tArgs:\n",
    "\t\tstrand (str): strand information\n",
    "\t\tstart (int): genomic start position regardless of strand\n",
    "\t\tend (int): genomic end position regardless of strand\n",
    "\t\tread_splicing (tuple): corrected read splicing\n",
    "\t\tchrand_ref_mutple_exon_trans (dict): reference multi-exon transcript\n",
    "\t\"\"\"\n",
    "\tif strand == '+':\n",
    "\t\trss_dis = abs(chrand_ref_mutple_exon_trans[read_splicing][0]-start)\n",
    "\t\tres_dis = abs(chrand_ref_mutple_exon_trans[read_splicing][1]-end)\n",
    "\telse:\n",
    "\t\trss_dis = abs(chrand_ref_mutple_exon_trans[read_splicing][0]-end)\n",
    "\t\tres_dis = abs(chrand_ref_mutple_exon_trans[read_splicing][1]-start)\n",
    "\t\n",
    "\treturn rss_dis, res_dis\n",
    "\n",
    "\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def multi_exon_read_correction(chrom, strand, name, full_block, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window, mis_intron_length, corExcept_dis, polya_dict):\n",
    "\t\"\"\"splicing junction correction and collaspsing for multi-exon read\n",
    "\n",
    "\tArgs:\n",
    "\t\tchrom (str): chromosome\n",
    "\t\tstrand (str): strand information\n",
    "\t\tname (str): read name\n",
    "\t\tfull_block (list): start, end position and all splicing site of multi-exon read (1 base, [1,20,40, 100])\n",
    "\t\tchrand_ref_exon (interlap data): exon from reference multi-exon transcript\n",
    "\t\tchrand_ref_junction (list): reference splicing junction\n",
    "\t\tchrand_ref_single_exon_trans (interlap data): reference single-exon transcript\n",
    "\t\tchrand_ref_mutple_exon_trans (dict): reference multi-exon transcript\n",
    "\t\tchrand_left_sj_set (list): reference left splicing site\n",
    "\t\tchrand_right_sj_set (list): reference right splicing site\n",
    "\t\tchrand_junction_dict (dict): splicing junction detected from long read\n",
    "\t\tsj_correction_window (int): tolerance distance for splicing site correction\n",
    "\t\tmis_intron_length (int): unintended small intron gap that should be filled\n",
    "\t\tcorExcept_dis (int, optional): edit distance to the reference splicing site\n",
    "\t\tpolya_dict (dict, optional): raw long reads Polyadenylation event\n",
    "\t\"\"\"\n",
    "\n",
    "\traw_splicing= tuple(full_block[1:-1])\n",
    "\tstart, end = full_block[0], full_block[-1]\n",
    "\tcorrected_read_splicing = []\n",
    "\ttag_dict={'reference_match':False, 'multi-exon':True, 'correct_site':[], 'merge_site':[], 'polya_signal':False , 'lowCredit_junction':{}, 'splicing_tag':[], 'rss_dis':None, 'res_dis':None}\n",
    "\n",
    "\t# polya event checking\n",
    "\ttry:\n",
    "\t\tif polya_dict[name]: tag_dict['polya_signal'] = True\n",
    "\texcept:\n",
    "\t\tpass\n",
    "\t\n",
    "\tif raw_splicing in chrand_ref_mutple_exon_trans: # uncorrected read_splicing matching with the reference\n",
    "\t\ttag_dict['reference_match'] = True\n",
    "\t\ttag_dict['splicing_tag'] = 'FSM'\n",
    "\t\tcorrected_read_splicing = raw_splicing\n",
    "\t\ttag_dict['rss_dis'], tag_dict['res_dis'] = RTS_refrence_distance(strand, start, end, corrected_read_splicing, chrand_ref_mutple_exon_trans)\n",
    "\n",
    "\telse: # splicing site correction\n",
    "\t\titered_raw_splicing = iter(raw_splicing)\n",
    "\t\tfor idx, (left_sj, right_sj) in enumerate(zip(itered_raw_splicing, itered_raw_splicing)):\n",
    "\n",
    "\t\t\t# check splicing coverage and motif in raw data\n",
    "\t\t\tsj_pos = full_block.index(left_sj)\n",
    "\t\t\ttmp_sj = (chrom,strand,left_sj,right_sj)\n",
    "\t\t\tjunction_coverage = chrand_junction_dict[tmp_sj][0]\n",
    "\t\t\tjunction_motif = chrand_junction_dict[tmp_sj][2]\n",
    "\t\t\t\n",
    "\t\t\tif left_sj in chrand_left_sj_set and right_sj in chrand_right_sj_set:\n",
    "\t\t\t\tif (left_sj, right_sj) in chrand_ref_junction:\n",
    "\t\t\t\t\ttag_dict['splicing_tag'].append('M')\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\ttag_dict['splicing_tag'].append('KC')\n",
    "\n",
    "\t\t\telif chrand_left_sj_set:\n",
    "\t\t\t\tleft_dis, left_ref_sj = loc_distance(chrand_left_sj_set, left_sj)\n",
    "\t\t\t\tright_dis, right_ref_sj = loc_distance(chrand_right_sj_set, right_sj)\n",
    "\n",
    "\t\t\t\t# do not correct the splicing site once the edit distance > sj_correction_window\n",
    "\t\t\t\tif left_dis > sj_correction_window: left_ref_sj = left_sj\n",
    "\t\t\t\tif right_dis > sj_correction_window: right_ref_sj = right_sj\n",
    "\n",
    "\t\t\t\t# correction exception, splicing junction with edit distance less than 4 for both sides\n",
    "\t\t\t\tif left_dis <= corExcept_dis and right_dis <= corExcept_dis and junction_coverage > 1 and junction_motif == 'canonical':\n",
    "\t\t\t\t\ttag_dict['splicing_tag'].append('EXC')\n",
    "\t\t\t\t\t\n",
    "\t\t\t\t# splicing site correction\n",
    "\t\t\t\telif left_dis <= sj_correction_window or right_dis <= sj_correction_window:\n",
    "\t\t\t\t\tif [left_sj, right_sj] == [left_ref_sj, right_ref_sj]:\n",
    "\t\t\t\t\t\tpass\n",
    "\t\t\t\t\telif full_block[sj_pos-1] < left_ref_sj < right_ref_sj < full_block[sj_pos+2]:\n",
    "\t\t\t\t\t\ttag_dict['correct_site'].append([left_sj, right_sj])\n",
    "\t\t\t\t\t\tleft_sj, right_sj = left_ref_sj, right_ref_sj\n",
    "\n",
    "\t\t\t\t\tif left_sj in chrand_left_sj_set and right_sj in chrand_right_sj_set:\n",
    "\t\t\t\t\t\tif (left_sj, right_sj) in chrand_ref_junction:\n",
    "\t\t\t\t\t\t\ttag_dict['splicing_tag'].append('CM')\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\ttag_dict['splicing_tag'].append('CKC')\n",
    "\n",
    "\t\t\t#checking unintended small intron overlap with reference exons\n",
    "\t\t\tif len(tag_dict['splicing_tag']) == idx and right_sj - left_sj <= mis_intron_length:\n",
    "\t\t\t\tif chrand_ref_exon:\n",
    "\t\t\t\t\toverlapped_ref_exon = tuple(chrand_ref_exon.find((left_sj, right_sj)))\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\toverlapped_ref_exon = ()\n",
    "\t\t\t\t# compare the unintended intron with the ref_exon from multi-exon transcripts\n",
    "\t\t\t\tif idx == 0 and overlapped_ref_exon:\n",
    "\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\tif (len(full_block) == 4 and exon[1] >= right_sj and exon[0] <= left_sj):\n",
    "\t\t\t\t\t\t\ttag_dict['merge_site'].append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\ttag_dict['multi-exon'] = False\n",
    "\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\telif (exon[1] == full_block[sj_pos+2] and exon[0] <= left_sj):\n",
    "\t\t\t\t\t\t\ttag_dict['merge_site'].append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\n",
    "\t\t\t\telif idx == int(len(raw_splicing)/2 - 1) and overlapped_ref_exon:\n",
    "\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\tif exon[1] >= right_sj and exon[0] == full_block[sj_pos-1]:\n",
    "\t\t\t\t\t\t\ttag_dict['merge_site'].append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\n",
    "\t\t\t\telif 0 < idx < int(len(raw_splicing)/2 -1) and overlapped_ref_exon:\n",
    "\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\tif exon[1] == full_block[sj_pos+2]  and exon[0] == full_block[sj_pos-1]:\n",
    "\t\t\t\t\t\t\ttag_dict['merge_site'].append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\n",
    "\t\t\t\t# compare the unintended intron with the exon from single-exon transcripts\n",
    "\t\t\t\telif not overlapped_ref_exon and len(full_block) == 4:\n",
    "\t\t\t\t\tif chrand_ref_single_exon_trans:\n",
    "\t\t\t\t\t\toverlapped_ref_exon = tuple(chrand_ref_single_exon_trans.find((left_sj, right_sj)))\n",
    "\t\t\t\t\n",
    "\t\t\t\t\tif overlapped_ref_exon:\n",
    "\t\t\t\t\t\tfor exon in overlapped_ref_exon:\n",
    "\t\t\t\t\t\t\tif exon[1] >= right_sj and exon[0] <= left_sj:\n",
    "\t\t\t\t\t\t\t\ttag_dict['merge_site'].append([full_block[sj_pos], full_block[sj_pos+1]])\n",
    "\t\t\t\t\t\t\t\ttag_dict['multi-exon'] = False\n",
    "\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\n",
    "\t\t\tcorrected_read_splicing.extend([left_sj, right_sj])\n",
    "\n",
    "\t\t\tif [full_block[sj_pos], full_block[sj_pos+1]] in tag_dict['merge_site']:\n",
    "\t\t\t\ttag_dict['splicing_tag'].append('UI')\n",
    "\t\t\t\tdel corrected_read_splicing[-2:]\n",
    "\n",
    "\t\t\telif len(tag_dict['splicing_tag']) == idx:\n",
    "\t\t\t\ttag_dict['splicing_tag'].append('NC')\n",
    "\t\t\t\tif junction_coverage == 1:\n",
    "\t\t\t\t\ttag_dict['lowCredit_junction'][idx+1] = junction_motif\n",
    "\n",
    "\t\tcorrected_read_splicing = tuple(corrected_read_splicing)\n",
    "\t\tif corrected_read_splicing in chrand_ref_mutple_exon_trans:\n",
    "\t\t\ttag_dict['reference_match'] = True\n",
    "\t\t\ttag_dict['rss_dis'], tag_dict['res_dis'] = RTS_refrence_distance(strand, start, end, corrected_read_splicing, chrand_ref_mutple_exon_trans)\n",
    "\t\n",
    "\treturn start, end, corrected_read_splicing, tag_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def multi_exon_read_collapse(read_id, start, end, corrected_read_splicing, tag_dict, rss_dis_lst, res_dis_lst, multi_exon_read, collapsed_idx):\n",
    "\t\"\"\"collapsing multi-exon read\n",
    "\t\"\"\"\n",
    "\tread_name = read_id.split('_', 1)[1]\n",
    "\tif tag_dict['lowCredit_junction']:\n",
    "\t\tpass\n",
    "\telif corrected_read_splicing:\n",
    "\t\tif corrected_read_splicing not in multi_exon_read:\n",
    "\t\t\tcollapsed_idx += 1\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing] = [[start], [end], [tag_dict['polya_signal']], 1, [read_name], tag_dict['reference_match'], f'collapsed.{collapsed_idx}']\n",
    "\t\telse:\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing][0].insert(0, start)\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing][1].insert(0, end)\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing][2].insert(0,tag_dict['polya_signal'])\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing][3] += 1\n",
    "\t\t\tmulti_exon_read[corrected_read_splicing][4].insert(0,read_name)\n",
    "\n",
    "\tif tag_dict['rss_dis']:\n",
    "\t\trss_dis_lst.append(tag_dict['rss_dis'])\n",
    "\t\tres_dis_lst.append(tag_dict['res_dis'])\n",
    "\n",
    "\treturn multi_exon_read, rss_dis_lst, res_dis_lst, collapsed_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def three_prime_exon_extraction(strand, multi_exon_read):\n",
    "\tthree_prime_exon = defaultdict(set)\n",
    "\tfor corrected_read_splicing, read_info in multi_exon_read.items():\n",
    "\t\tif strand == '+':\n",
    "\t\t\tmax_end = max(read_info[1])\n",
    "\t\t\texon = (corrected_read_splicing[-1], max_end)\n",
    "\t\telse:\n",
    "\t\t\tmax_end = max(read_info[0])\n",
    "\t\t\texon = (max_end, corrected_read_splicing[0])\n",
    "\t\tthree_prime_exon.add(exon)\n",
    "\t\n",
    "\treturn three_prime_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def coco_operation (chrom, strand, chrand_processed_read, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window = 40, mis_intron_length = 150, corExcept_dis = 0, polya_dict=None, collapsed_idx = 0):\n",
    "\t\"\"\"main function for correcting splicing junction and collpasing reads\n",
    "\t\"\"\"\n",
    "\tcorrected_read = defaultdict(dict)\n",
    "\tcorrection_log = defaultdict(dict)\n",
    "\tsingle_exon_read = defaultdict(dict)\n",
    "\tmulti_exon_read = defaultdict(dict)\n",
    "\trss_dis_lst = res_dis_lst = []\n",
    "\n",
    "\tfor read_id, full_block in tqdm(chrand_processed_read.items(), desc = f'{strftime(\"%Y-%m-%d %H:%M:%S\")}: Collapsing raw reads from {chrom} {strand}'):\n",
    "\t\t\n",
    "\t\t# single-exon read collapsing\n",
    "\t\tif len(full_block) == 2: \n",
    "\t\t\tsingle_exon_read = single_exon_read_collapse(full_block, chrand_ref_exon, sj_correction_window, single_exon_read)\n",
    "\t\t\tcorrected_read[read_id] = full_block\n",
    "\t\t# multi-exon read correction and collapsing\n",
    "\t\telse:\n",
    "\t\t\tstart, end, corrected_read_splicing, tag_dict = multi_exon_read_correction(chrom, strand, read_id, full_block, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window, mis_intron_length, corExcept_dis, polya_dict)\n",
    "\t\t\t\n",
    "\t\t\tmulti_exon_read, rss_dis_lst, res_dis_lst, collapsed_idx = multi_exon_read_collapse(read_id, start, end, corrected_read_splicing, tag_dict, rss_dis_lst, res_dis_lst, multi_exon_read, collapsed_idx)\n",
    "\t\t\tif corrected_read_splicing:\n",
    "\t\t\t\tcorrected_read[read_id] = [start, *list(corrected_read_splicing), end]\n",
    "\t\t\telse:\n",
    "\t\t\t\tcorrected_read[read_id] = [start, end]\n",
    "\t\t\tcorrection_log[read_id] = tag_dict\n",
    "\t\t\t\t\n",
    "\treturn chrom, strand, single_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "def splicing_to_bed_block(chrom, strand, name, full_block):\n",
    "\tstart, end = full_block[0], full_block[-1]\n",
    "\tfull_block = iter(full_block)\n",
    "\tblock_sizes = []\n",
    "\tblock_starts = []\n",
    "\tfor left_end, right_end in zip(full_block, full_block):\n",
    "\t\tblock_starts.append(left_end - start)\n",
    "\t\tblock_sizes.append(right_end - left_end + 1)\n",
    "\tblock_count = len(block_sizes)\n",
    "\tblock_sizes = ','.join([str(i) for i in block_sizes])\n",
    "\tblock_starts = ','.join([str(i) for i in block_starts])\n",
    "\tbed_block = [chrom, start-1, end, name, '-', strand, start-1, end, '255,0,0', block_count, block_sizes, block_starts]\n",
    "\tbed_block = '\\t'.join([str(i) for i in bed_block])\n",
    "\n",
    "\treturn bed_block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "class CoCoWrapper:\n",
    "\tdef __init__(self, thread, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir, sj_correction_window, polya_dict, mis_intron_length, corExcept_dis=4):\n",
    "\t\tself.thread = thread\n",
    "\t\tself.processed_read = processed_read\n",
    "\t\tself.ref_exon = ref_exon\n",
    "\t\tself.ref_junction = ref_junction\n",
    "\t\tself.ref_single_exon_trans = ref_single_exon_trans\n",
    "\t\tself.ref_mutple_exon_trans = ref_mutple_exon_trans\n",
    "\t\tself.left_sj_set = left_sj_set\n",
    "\t\tself.right_sj_set = right_sj_set\n",
    "\t\tself.junction_dict = junction_dict\n",
    "\t\tself.sj_correction_window = sj_correction_window\n",
    "\t\tself.tmp_dir = tmp_dir\n",
    "\t\tself.polya_dict = polya_dict\n",
    "\t\tself.mis_intron_length = mis_intron_length\n",
    "\t\tself.corExcept_dis = corExcept_dis\n",
    "\t\t\n",
    "\n",
    "\tdef job_compute(self):\n",
    "\t\tp = Pool(processes = self.thread)\n",
    "\t\tresult = []\n",
    "\t\tfor branch in self.processed_read:\n",
    "\t\t\tchrom, strand = branch\n",
    "\t\t\tchrand_processed_read = self.processed_read[branch]\n",
    "\t\t\tchrand_ref_exon = self.ref_exon[branch]\n",
    "\t\t\tchrand_ref_junction = self.ref_junction[branch]\n",
    "\t\t\tchrand_ref_single_exon_trans = self.ref_single_exon_trans[branch]\n",
    "\t\t\tchrand_ref_mutple_exon_trans = self.ref_mutple_exon_trans[branch]\n",
    "\t\t\tchrand_left_sj_set = self.left_sj_set[branch]\n",
    "\t\t\tchrand_right_sj_set = self.right_sj_set[branch]\n",
    "\t\t\tchrand_junction_dict = self.junction_dict[branch]\n",
    "\t\t\tresult.append(p.apply_async(coco_operation, (chrom, strand, chrand_processed_read, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, self.sj_correction_window, self.mis_intron_length, self.corExcept_dis, self.polya_dict,)))\n",
    "\n",
    "\t\tp.close()\n",
    "\t\tp.join()\n",
    "\n",
    "\t\treturn result\n",
    "\n",
    "\tdef result_collection(self):\n",
    "\t\tcollected_single_exon_read = Vividict()\n",
    "\t\tcollected_multi_exon_read = Vividict()\n",
    "\t\tcollected_rss = []\n",
    "\t\tcollected_res = []\n",
    "\t\tpath_to_log = f'{self.tmp_dir}/read_correction.log'\n",
    "\t\tpath_to_corrected_bed = f'{self.tmp_dir}/Corrected_reads.bed'\n",
    "\n",
    "\t\tresult = self.job_compute()\n",
    "\t\twith open(path_to_log, 'w') as flog, open(path_to_corrected_bed, 'w') as fbed:\n",
    "\t\t\tfor res in result:\n",
    "\t\t\t\tchrom, strand, single_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst = res.get()\n",
    "\t\t\t\tcollected_single_exon_read[(chrom, strand)] = single_exon_read\n",
    "\t\t\t\tcollected_multi_exon_read[(chrom, strand)] = multi_exon_read\n",
    "\t\t\t\tcollected_rss.extend(rss_dis_lst)\n",
    "\t\t\t\tcollected_res.extend(res_dis_lst)\n",
    "\t\t\t\tfor read_id, tag_dict in correction_log.items():\n",
    "\t\t\t\t\tread_name = read_id.split('_', 1)[1]\n",
    "\t\t\t\t\tattributes = '\\t'.join('{}: {}'.format(key, value) for key, value in tag_dict.items())\n",
    "\t\t\t\t\tflog.write(f'{read_name}\\t{attributes}\\n')\n",
    "\t\t\t\t\n",
    "\t\t\t\tfor read_id, full_block in corrected_read.items():\n",
    "\t\t\t\t\tread_name = read_id.split('_', 1)[1]\n",
    "\t\t\t\t\tbed_block = splicing_to_bed_block(chrom, strand, read_name, full_block)\n",
    "\t\t\t\t\tfbed.write(f'{bed_block}\\n')\n",
    "\t\treturn collected_single_exon_read, collected_multi_exon_read, collected_rss, collected_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chrand_read_list1 = {}\n",
    "# chrand_read_list1['adcc854a-2c55-432c-9d65-11ca1d8c9eb4'] = chrand_read_list['adcc854a-2c55-432c-9d65-11ca1d8c9eb4']\n",
    "# # chrand_read_list1['adcc854a-2c55-432c-9d65-11ca1d8c9eb4'] = [167500,167519, 167584, 167610]\n",
    "\n",
    "\n",
    "# for (chrom, strand) in processed_read:\n",
    "# \tchrand_read_list = processed_read[(chrom, strand)]\n",
    "# \tchrand_ref_exon = ref_exon[(chrom, strand)]\n",
    "# \tchrand_ref_junction = ref_junction[(chrom, strand)]\n",
    "# \tchrand_ref_single_exon_trans = ref_single_exon_trans[(chrom, strand)]\n",
    "# \tchrand_ref_mutple_exon_trans = ref_mutple_exon_trans[(chrom,strand)]\n",
    "# \tchrand_left_sj_set = left_sj_set[(chrom,strand)]\n",
    "# \tchrand_right_sj_set = right_sj_set[(chrom,strand)]\n",
    "# \tchrand_junction_dict = junction_dict[(chrom,strand)]\n",
    "# \tsj_correction_window = 40\n",
    "# \tmis_intron_length = 150\n",
    "# \tcorExcept_dis=4\n",
    "# \tpolya_dict=None\n",
    "# \tsingle_exon_read = {}\n",
    "# \ttmp_dir='./'\n",
    "# \t# with open ('test.out', 'w') as fw:\n",
    "# \t# \tfor read, splicing in tqdm(chrand_read_list.items(), desc = f'{strftime(\"%Y-%m-%d %H:%M:%S\")}: Collapsing raw reads from {chrom} {strand}'):\n",
    "# \t# \t\tif len(splicing) == 2:\n",
    "# \t# \t\t\tsingle_exon_read_collapse(splicing, chrand_ref_exon, 40, single_exon_ead)\n",
    "# \t# \t\tif len(splicing) > 2:\n",
    "\n",
    "# \tsingle_exon_read, multi_exon_read, corrected_read, correction_log, rss_dis_lst, res_dis_lst = coco_operation(chrom, strand, chrand_read_list, chrand_ref_exon, chrand_ref_junction, chrand_ref_single_exon_trans, chrand_ref_mutple_exon_trans, chrand_left_sj_set, chrand_right_sj_set, chrand_junction_dict, sj_correction_window)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "collected_single_exon_read, collected_multi_exon_read = CoCoWrapper(32, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir='.', sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=None).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
