{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: tailfinder.html\n",
    "title: tailFinder\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp tailFinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "import warnings\n",
    "import pandas as pd\n",
    "from collections import defaultdict, Counter\n",
    "from interlap import InterLap\n",
    "from joblib import Parallel, delayed\n",
    "from time import strftime\n",
    "from tqdm import tqdm\n",
    "from sklearn.metrics import silhouette_score\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from LAFITE.utils import Vividict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gtf = '/expt/zjzace/Nanopore_subcellular/Reference/gencode.v38.primary_assembly.annotation.sorted.gtf'\n",
    "# ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefAnnotationExtraction(gtf).annotation_sorting()\n",
    "# bed = '/expt/zjzace/Nanopore_subcellular/Analysis/Assembly/LAFITE/GLRA_tmp/A549_Cyto_LAFITE_tmp/bam.bed'\n",
    "# fa = '/NFS/mnemosyne2/expt/zjzace/GenomeRef/GRCh38.primary_assembly.genome.fa'\n",
    "# junction_dict, processed_read = read_grouping(bed, fa)\n",
    "\n",
    "# polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/Analysis/Nanopolish/A549_Cyto_PolyA.res')\n",
    "# collected_single_exon_read, collected_multi_exon_read = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir='.', sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=polya_dict).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# gtf = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta-annotation_C_170612a.gtf'\n",
    "# bed = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.sorted.bed'\n",
    "# fa = '/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/Raw_data/SIRV_isoforms_multi-fasta_170612a.fasta'\n",
    "# junction_dict, processed_read = read_grouping(bed, fa)\n",
    "# ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, tss_dict = RefAnnotationExtraction(gtf).annotation_sorting()\n",
    "\n",
    "# # polya_dict = polya_signal_import('/expt/zjzace/Nanopore_subcellular/SIRV/SIRV_Set1/bam/SRR6058584.polya.res')\n",
    "# polya_dict = PolyAFinder(processed_read, fa, '/home/zjzace/software/SQANTI3-4.1/data/polyA_motifs/mouse_and_human.polyA_motif.txt').polya_estimation()\n",
    "# collected_single_exon_read, collected_multi_exon_read, tss, rss = CoCoWrapper(16, processed_read, ref_exon, ref_junction, ref_single_exon_trans, ref_mutple_exon_trans, left_sj_set, right_sj_set, junction_dict, tmp_dir='.', sj_correction_window=40, mis_intron_length = 150, corExcept_dis=4, polya_dict=polya_dict).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class AlternativeTerminalFinder:\n",
    "\tdef __init__(self, chrom, strand, corrected_read_splicing, read_info, min_count_tss_tes, max_sil = 0):\n",
    "\n",
    "\t\tself.chrom = chrom\n",
    "\t\tself.strand = strand\n",
    "\t\tself.corrected_read_splicing = corrected_read_splicing\n",
    "\t\tself.polya_lst = read_info[2]\n",
    "\t\tself.count = read_info[3]\n",
    "\t\tself.fsm = read_info[4]\n",
    "\t\tself.collapsed_ID = read_info[5]\n",
    "\t\t# print(read_info,len(read_info))\n",
    "\t\tself.reference_id = list(set(read_info[6]))\n",
    "\t\tself.min_count_tss_tes = min_count_tss_tes\n",
    "\t\tself.max_sil = max_sil\n",
    "\t\tself.polya_count = sum(self.polya_lst)\n",
    "\t\tif self.strand == '+':\n",
    "\t\t\tself.rss_lst = read_info[0]\n",
    "\t\t\tself.res_lst = read_info[1]\n",
    "\t\telse:\n",
    "\t\t\tself.rss_lst = read_info[1]\n",
    "\t\t\tself.res_lst = read_info[0]\n",
    "\n",
    "\tdef optimal_k(self, end_list):\n",
    "\t\tmax_sil = self.max_sil\n",
    "\t\tk_optimal = 1\n",
    "\t\tdf = pd.DataFrame(end_list, columns=['tts'])\n",
    "\t\twith warnings.catch_warnings(record=True) as w:\n",
    "\t\t\twarnings.filterwarnings(\"error\")\n",
    "\t\t\tfor k in range(2, 5):\n",
    "\t\t\t\ttry:\n",
    "\t\t\t\t\tgmm = GaussianMixture(n_components=k,random_state=0).fit(df)\n",
    "\t\t\t\t\tlabels = gmm.predict(df)\n",
    "\t\t\t\t\tcurr_sil = silhouette_score(df, labels, metric='euclidean')\n",
    "\t\t\t\t\tif max_sil < curr_sil:\n",
    "\t\t\t\t\t\tmax_sil = curr_sil\n",
    "\t\t\t\t\t\tk_optimal = k\n",
    "\t\t\t\texcept:\n",
    "\t\t\t\t\tk_optimal = k-1\n",
    "\t\t\t\t\tbreak\n",
    "\t\treturn df, k_optimal\n",
    "\n",
    "\tdef terminal_cluster(self):\n",
    "\t\toutlist = [self.chrom, self.strand, self.corrected_read_splicing, self.fsm, self.count, self.polya_count, self.collapsed_ID,self.reference_id]\n",
    "\t\tfor idx, tail_lst in enumerate([self.rss_lst, self.res_lst]):\n",
    "\n",
    "\t\t\tif idx == 0:\n",
    "\t\t\t\tmin_dis = 50\n",
    "\t\t\t\tpolya_lst = []\n",
    "\t\t\telse:\n",
    "\t\t\t\tmin_dis = 24\n",
    "\t\t\t\tpolya_lst = self.polya_lst\n",
    "\n",
    "\t\t\tif len(tail_lst) == 1:\n",
    "\t\t\t\tif idx == 0:\n",
    "\t\t\t\t\toutlist.extend([tail_lst[0],[tail_lst[0]]])\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tpolya_tag = True if polya_lst[0] else False\n",
    "\t\t\t\t\toutlist.extend([tail_lst[0], [tail_lst[0]], polya_tag])\n",
    "\t\t\telse:\n",
    "\t\t\t\tapa_dict = {}\n",
    "\t\t\t\tpolya_tag = False\n",
    "\t\t\t\tdf, k_optimal = self.optimal_k(tail_lst)\n",
    "\t\t\t\tgmm = GaussianMixture(n_components=k_optimal,random_state=0).fit(df)\n",
    "\t\t\t\tlabels = gmm.predict(df)\n",
    "\t\t\t\tdf['labels'] = labels\n",
    "\t\t\t\tclusters = {k:v for k, v in Counter(df['labels']).items() if v > 2}\n",
    "\t\t\t\tclusters = dict(sorted(clusters.items(), key=lambda e: e[1], reverse=True))\n",
    "\n",
    "\t\t\t\tif sum(polya_lst)/(self.count) >=0.4:\n",
    "\t\t\t\t\tdf['polya'] = polya_lst\n",
    "\t\t\t\t\tfor key, value in clusters.items():\n",
    "\t\t\t\t\t\tif df[(df['labels']==key)&(df['polya']==True)].shape[0]/df[df['labels']==key].shape[0] >= 0.2:\n",
    "\t\t\t\t\t\t\tif apa_dict:\n",
    "\t\t\t\t\t\t\t\tapa_count = df[df['labels']==key]['tts'].value_counts().tolist()[0]\n",
    "\t\t\t\t\t\t\t\tif apa_count >= self.min_count_tss_tes:\n",
    "\t\t\t\t\t\t\t\t\tapa_site = df[df['labels']==key]['tts'].value_counts().index[0]\n",
    "\t\t\t\t\t\t\t\t\tfor tmp_site,tmp_count in apa_dict.copy().items():\n",
    "\t\t\t\t\t\t\t\t\t\tif abs(tmp_site-apa_site) < min_dis and apa_count > tmp_count[0]:\n",
    "\t\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\t\t\t\t\t\tapa_dict.pop(tmp_site)\n",
    "\t\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\t\telif abs(tmp_site-apa_site) < min_dis and apa_count == tmp_count[0] and value > tmp_count[1]:\n",
    "\t\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\t\t\t\t\t\tapa_dict.pop(tmp_site)\n",
    "\t\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\t\telif abs(tmp_site-apa_site) < min_dis:\n",
    "\t\t\t\t\t\t\t\t\t\t\tpass\n",
    "\t\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\t\tapa_count = df[df['labels']==key]['tts'].value_counts().tolist()[0]\n",
    "\t\t\t\t\t\t\t\tif apa_count >= self.min_count_tss_tes:\n",
    "\t\t\t\t\t\t\t\t\tapa_site = df[df['labels']==key]['tts'].value_counts().index[0]\n",
    "\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\tif apa_dict:\n",
    "\t\t\t\t\t\t\tpolya_tag = True\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tfor key, value in clusters.items():\n",
    "\t\t\t\t\t\tif apa_dict:\n",
    "\t\t\t\t\t\t\tapa_count = df[df['labels']==key]['tts'].value_counts().tolist()[0]\n",
    "\t\t\t\t\t\t\tif apa_count >= self.min_count_tss_tes:\n",
    "\t\t\t\t\t\t\t\tapa_site = df[df['labels']==key]['tts'].value_counts().index[0]\n",
    "\t\t\t\t\t\t\t\tfor tmp_site,tmp_count in apa_dict.copy().items():\n",
    "\t\t\t\t\t\t\t\t\tif abs(tmp_site-apa_site) < min_dis and apa_count > tmp_count[0]:\n",
    "\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\t\t\t\t\tapa_dict.pop(tmp_site)\n",
    "\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\telif abs(tmp_site-apa_site) < min_dis and apa_count == tmp_count[0] and value > tmp_count[1]:\n",
    "\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\t\t\t\t\tapa_dict.pop(tmp_site)\n",
    "\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\telif abs(tmp_site-apa_site) < min_dis:\n",
    "\t\t\t\t\t\t\t\t\t\tpass\n",
    "\t\t\t\t\t\t\t\t\t\tbreak\n",
    "\t\t\t\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tapa_count = df[df['labels']==key]['tts'].value_counts().tolist()[0]\n",
    "\t\t\t\t\t\t\tif apa_count >= self.min_count_tss_tes:\n",
    "\t\t\t\t\t\t\t\tapa_site = df[df['labels']==key]['tts'].value_counts().index[0]\n",
    "\t\t\t\t\t\t\t\tapa_dict[apa_site] = [apa_count, value]\n",
    "\n",
    "\t\t\t\tif apa_dict:\n",
    "\t\t\t\t\tapa_site = list(apa_dict.keys())\n",
    "\t\t\t\t\tend = apa_site[0]\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\tif sum(polya_lst)/(self.count) >=0.4:\n",
    "\t\t\t\t\t\tif df[df['polya'] == True].shape[0]/df.shape[0] >= 0.2:\n",
    "\t\t\t\t\t\t\tpolya_tag =True\n",
    "\t\t\t\t\t\tdf = df[['tts','polya']].value_counts().reset_index(name='counts')\n",
    "\t\t\t\t\t\tdf = df.pivot(index='tts', columns='polya', values='counts')\n",
    "\t\t\t\t\t\tdf = df.reset_index(level=['tts'])\n",
    "\t\t\t\t\t\tdf = df.fillna(0)\n",
    "\t\t\t\t\t\tif False not in df:\n",
    "\t\t\t\t\t\t\tdf[False] = 0\n",
    "\t\t\t\t\t\tdf['ratio'] = df[True]/(df[False]+df[True])\n",
    "\t\t\t\t\t\tif self.strand == '+':\n",
    "\t\t\t\t\t\t\tdf = df.sort_values(['ratio',True, 'tts', False], ascending=[False,False,False,True])\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tdf = df.sort_values(['ratio',True, 'tts', False], ascending=[False,False,True,True])\n",
    "\t\t\t\t\telse:\n",
    "\t\t\t\t\t\tdf = df['tts'].value_counts().reset_index()\n",
    "\t\t\t\t\t\tdf.columns = ['tts', 'counts']\n",
    "\t\t\t\t\t\tif (self.strand == '+' and idx == 1) or (self.strand == '-' and idx == 0):\n",
    "\t\t\t\t\t\t\tdf = df.sort_values(['counts','tts'], ascending=[False,False])\n",
    "\t\t\t\t\t\telif (self.strand == '+' and idx == 0) or (self.strand == '-' and idx == 1):\n",
    "\t\t\t\t\t\t\tdf = df.sort_values(['counts','tts'], ascending=[False,True])\n",
    "\t\t\t\t\tend = df['tts'].iloc[0]\n",
    "\t\t\t\t\tapa_site = [end]\n",
    "\t\t\t\tif idx == 0:\n",
    "\t\t\t\t\toutlist.extend([end, apa_site])\n",
    "\t\t\t\telse:\n",
    "\t\t\t\t\toutlist.extend([end, apa_site, polya_tag])\n",
    "\t\treturn outlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exports\n",
    "\n",
    "class TailFinderWrapper:\n",
    "\tdef __init__(self, collected_multi_exon_read, min_count_tss_tes, thread):\n",
    "\t\tself.collected_multi_exon_read = collected_multi_exon_read\n",
    "\t\tself.min_count_tss_tes = min_count_tss_tes\n",
    "\t\tself.thread = thread\n",
    "\n",
    "\tdef job_precompute(self):\n",
    "\t\tprecompute_list = []\n",
    "\t\tfor (chrom, strand), read_dict in self.collected_multi_exon_read.items():\n",
    "\t\t\tfor corrected_read_splicing, read_info in read_dict.items():\n",
    "\t\t\t\tprecompute_list.append(AlternativeTerminalFinder(chrom, strand, corrected_read_splicing, read_info, self.min_count_tss_tes))\n",
    "\t\t\n",
    "\t\treturn precompute_list\n",
    "\n",
    "\tdef run(self):\n",
    "\t\tprecompute_list = self.job_precompute()\n",
    "\t\twith Parallel(n_jobs = self.thread) as parallel:\n",
    "\t\t\tresults_lst = parallel(delayed(lambda x:x.terminal_cluster())(job) for job in tqdm(precompute_list, desc = f'{strftime(\"%Y-%m-%d %H:%M:%S\")}: Calculating pupative TSS and TES for collapsed read'))\n",
    "\t\t\n",
    "\t\treturn results_lst\n",
    "\t\n",
    "\tdef return_extremum(self, strand, entry):\n",
    "\t\t\n",
    "\t\tif strand == '+':\n",
    "\t\t\textremum = max(entry)\n",
    "\t\telse:\n",
    "\t\t\textremum = min(entry)\n",
    "\t\treturn extremum\n",
    "\n",
    "\tdef result_collection(self):\n",
    "\t\tresults_lst = self.run()\n",
    "\t\tprocessed_collected_multi_exon_read = Vividict()\n",
    "\t\tthree_prime_exon = defaultdict(set)\n",
    "\t\tfor result in results_lst:\n",
    "\t\t\tchrom, strand, corrected_read_splicing, fsm, total_count, polya_count, collapsed_ID, reference_id, start, as_site, end, apa_site, polya_tag = result\n",
    "\t\t\textremum = self.return_extremum(strand, apa_site)\n",
    "\t\t\tif strand == '+':\n",
    "\t\t\t\tthree_prime_exon[(chrom, strand)].add((corrected_read_splicing[-1],extremum))\n",
    "\t\t\telse:\n",
    "\t\t\t\tthree_prime_exon[(chrom, strand)].add((extremum, corrected_read_splicing[0]))\n",
    "\t\t\tprocessed_collected_multi_exon_read[(chrom, strand)][corrected_read_splicing] = [start, end, total_count, polya_count, fsm, polya_tag, as_site, apa_site, collapsed_ID, reference_id]\n",
    "\t\t\n",
    "\t\t# sort by read exon number\n",
    "\n",
    "\t\tfor branch in processed_collected_multi_exon_read:\n",
    "\t\t\ttmp_dict = {}\n",
    "\t\t\tfor k in sorted(processed_collected_multi_exon_read[branch], key=len, reverse=True):\n",
    "\t\t\t\ttmp_dict[k] = processed_collected_multi_exon_read[branch][k]\n",
    "\t\t\tprocessed_collected_multi_exon_read[branch] = tmp_dict\n",
    "\n",
    "\t\t# convert three_prime_exon dict to interlap structure\n",
    "\t\tfor i in three_prime_exon:\n",
    "\t\t\tt = list(three_prime_exon[i])\n",
    "\t\t\tthree_prime_exon[i] = InterLap()\n",
    "\t\t\tthree_prime_exon[i].update(t)\n",
    "\n",
    "\t\treturn processed_collected_multi_exon_read, three_prime_exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-04-08 17:17:10: Calculating pupative TSS and TES for collapsed read: 100%|██████████| 4888/4888 [00:09<00:00, 495.06it/s] \n"
     ]
    }
   ],
   "source": [
    "processed_collected_multi_exon_read, three_prime_exon = TailFinderWrapper(collected_multi_exon_read, 3, 16).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_ref(ref_mutple_exon_trans):\n",
    "\treversed_ref_mutple_exon_trans = Vividict()\n",
    "\tfor (chrom, strand), isoforms in ref_mutple_exon_trans.items():\n",
    "\t\tfor isoform in isoforms:\n",
    "\t\t\tread_info = isoforms[isoform]\n",
    "\t\t\tif strand == '-':\n",
    "\t\t\t\tisoform = tuple(reversed(isoform))\n",
    "\t\t\treversed_ref_mutple_exon_trans[(chrom, strand)][isoform] = read_info\n",
    "\n",
    "\treturn reversed_ref_mutple_exon_trans\n",
    "\n",
    "ref_mutple_exon_trans = reverse_ref(ref_mutple_exon_trans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "collected_refined_isoforms = RefineWrapper(processed_collected_multi_exon_read, collected_single_exon_read, ref_mutple_exon_trans, ref_single_exon_trans, three_prime_exon, tss_dict, 30, 30, 3, 4, 50, 16).result_collection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
