{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "output-file: index.html\n",
    "title: LAFITE\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65c69e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# | include: false\n",
    "from LAFITE import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b54e66e",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "LAFITE is designated to identify high-consensus full-length isoforms from Nanopore Direct RNA-seq data. LAFITE combines multiple features from reference annotation and DRS reads (TSS, TES, splicing junction, and read polyadenylation event) and is more sensitive to Low-abundance transcripts.\n",
    "\n",
    "\n",
    "\n",
    "Prerequisites\n",
    "-------------\n",
    "* [bedtools](https://github.com/arq5x/bedtools2)\n",
    "* [Minimap2](https://github.com/lh3/minimap2)\n",
    "* [nanopolish](https://github.com/jts/nanopolish)\n",
    "* [samtools](http://www.htslib.org)\n",
    "* Python 3.7/3.8/3.9\n",
    "\n",
    "Installation\n",
    "------------\n",
    "To avoid potential conflicts, we recommend running LAFITE in a conda environment.\n",
    "```\n",
    "conda create -n LAFITE_env -c bioconda python=3.7 bedtools\n",
    "conda activate LAFITE_env\n",
    "pip install git+https://github.com/TF-Chan-Lab/LAFITE\n",
    "```\n",
    "\n",
    "Usage\n",
    "-----\n",
    "1. **Run minimap2 and samtools to generate alignment file in bam format**\n",
    "\t```\n",
    "\tminimap2 -ax splice -u f -k 14 -G 500000 --secondary=no REFERENCE_FA FASTQ > ALIGNMENT_SAM\n",
    "\tsamtools view -bS ALIGNMENT_SAM|samtools sort - > ALIGNMENT_BAM\n",
    "\t```\n",
    "\tLAFITE also supports other splicing-aware long read alignment tools.\n",
    "2. **Run Nanopolish polya to generate read polyadenylation result (optional but recommend)**  \n",
    "Current long-read sequencing technologies (Nanopore cDNA/DRS or PacBio Iso-Seq) are all designed to capture RNA molecules with poly(A) tail. However, RNA fragmentation and pore blocking may bring a considerable part of truncated reads which will interfere downstream analysis. Therefore, LAFITE utilizes the read polyadenylation status reported by Nanopolish to filter reads that have completed the sequencing process.  \n",
    "\n",
    "   ```\n",
    "   nanopolish index -d PATH_TO_FAST5 -s GUPPY_SEQUENCING_SUMMARY FASTQ\n",
    "   nanopolish polya -t NUM_OF_THREADS -r FASTQ -b ALIGNMENT_BAM -g REFERENCE_FA > Nanopolish_PolyA_RES\n",
    "   ```\n",
    "\tLAFITE also provides an alternative approach to estimate read polyadenylation status by scanning any poly(A) motifs that existed at the read 3'-end.  \n",
    "\n",
    "1. **Run LAFITE**  \n",
    "\t```\n",
    "\tusage: lafite [-h] -b BAM [-B BEDTOOLS] -g GTF -f GENOME -o OUTPUT\n",
    "              [-n MIN_COUNT_TSS_TES] [-i MIS_INTRON_LENGTH]\n",
    "              [-c MIN_NOVEL_TRANS_COUNT] [-s MIN_SINGLE_EXON_COVERAGE]\n",
    "              [-l MIN_SINGLE_EXON_LEN] [-L LABEL] [-p POLYA]\n",
    "              [-m POLYA_MOTIF_FILE] [-r RELATIVE_ABUNDANCE_THRESHOLD]\n",
    "              [-j SHORT_SJ_TAB] [-w SJ_CORRECTION_WINDOW] [--no_full_cleanup]\n",
    "              [-t THREAD] [-T TSS_PEAK] [-d TSS_CUTOFF]\n",
    "\n",
    "\tLow-abundance Aware Full-length Isoform clusTEr\n",
    "\n",
    "\toptional arguments:\n",
    "\t  -h, --help            show this help message and exit\n",
    "\t  -b BAM                path to the alignment file in bam format\n",
    "\t  -B BEDTOOLS           path to the executable bedtools\n",
    "\t  -g GTF                path to the reference gene annotation in GTF format\n",
    "\t  -f GENOME             path to the reference genome fasta\n",
    "\t  -o OUTPUT             path to the output file\n",
    "\t  -n MIN_COUNT_TSS_TES  minimum number of reads supporting a alternative TSS or TES, default: 3\n",
    "\t  -i MIS_INTRON_LENGTH  length cutoff for correcting unexpected small intron, default: 150\n",
    "\t  -c MIN_NOVEL_TRANS_COUNT\n",
    "\t                        minimum occurrences required for a isoform from novel loci, default: 3\n",
    "\t  -s MIN_SINGLE_EXON_COVERAGE\n",
    "\t                        minimum read coverage required for a novel single-exon transcript, default: 4\n",
    "\t  -l MIN_SINGLE_EXON_LEN\n",
    "\t                        minimum length for single-exon transcript, default: 100\n",
    "\t  -L LABEL              name prefix for output transcripts, default: LAFT\n",
    "\t  -p POLYA              path to the file contains read Polyadenylation event\n",
    "\t  -m POLYA_MOTIF_FILE   path to the polya motif file\n",
    "\t  -r RELATIVE_ABUNDANCE_THRESHOLD\n",
    "\t                        minimum abundance of the predicted multi-exon transcripts as a fraction of the\n",
    "\t\t\t\t\t\t\ttotal transcript assembled at a given locus, default: 0.01\n",
    "\t  -j SHORT_SJ_TAB       path to the short read splice junction file\n",
    "\t  -w SJ_CORRECTION_WINDOW\n",
    "\t                        edit distance to reference splicing site for splicing correction, default: 40\n",
    "\t  --no_full_cleanup     keep all intermediate files\n",
    "\t  -t THREAD             number of the threads, default: 4\n",
    "\t  -T TSS_PEAK           path to the TSS peak file\n",
    "\t  -d TSS_CUTOFF         minimum TSS distance for a transcript to be considered as a novel transcript\n",
    "\t```\n",
    "- LAFITE can run with the following arguments:\n",
    "   ```\n",
    "   lafite -b ALIGNMENT_BAM -g REFERENCE_GTF -f REFERENCE_FA -o OUTPUT_GTF -t NUM_OF_THREADS -p Nanopolish_PolyA_RES\n",
    "   ```\n",
    "- LAFITE can also run without the result from *nanoplish polya*. Then, a Poly(A) motif list must be provided for the corresponding species.  \n",
    "   We have provided the Poly(A) motif list for human and mouse retrieved from [*Tian* et al.](https://academic.oup.com/nar/article/33/1/201/2401035) .\n",
    "   \n",
    "   ```\n",
    "   lafite -b ALIGNMENT_BAM -g REFERENCE_GTF -f REFERENCE_FA -o OUTPUT_GTF -t NUM_OF_THREADS -m POLYA_MOTIFS_OF_SPECIES\n",
    "   ```\n",
    "- LAFITE accepts the TSS peaks from 5'-end CAGE data for identifying high-confidence TSSs. Users can prepare the TSS data in the following format where:\n",
    "  - The first column is the chromosome name\n",
    "  - The second column is the 0-based start position of the TSS peak\n",
    "  - The third column is the 1-based end position of the TSS peak\n",
    "  - The fourth column is the strand information  \n",
    "- LAFITE also accepts the splicing junctions from Illumina short read RNA-seq data to proof the long reads. LAFITE supports the SJ.out.tab from STAR aligner. Users can also prepare the splicing junctions in the following format where:\n",
    "  - The first column is the chromosome name\n",
    "  - The second column is the 0-based start position of the splicing junction\n",
    "  - The third column is the 1-based end position of the splicing junction\n",
    "  - The fourth column is the strand information  \n",
    "\n",
    "Development\n",
    "-----------\n",
    "\n",
    "LAFITE was developed following the [fastai/nbdev](https://github.com/fastai/nbdev) framework.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "016319be",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
